File: simulator/project-java/common-kafka/src/main/java/common/kafka/KafkaDispatcher.java
package common.kafka;

import java.io.Closeable;
import java.util.Properties;
import java.util.concurrent.ExecutionException;

import org.apache.kafka.clients.producer.Callback;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.apache.kafka.common.serialization.StringSerializer;

public class KafkaDispatcher<T> implements Closeable {
  private final KafkaProducer<String, T> producer;

  public KafkaDispatcher() {
    this.producer = new KafkaProducer<>(properties());
  }

  public void send(String topic, String key, T value) throws InterruptedException, ExecutionException {
    var record = new ProducerRecord<>(topic, key, value);
    Callback callback = (RecordMetadata metadata, Exception e) -> {
      if (e != null) {
        e.printStackTrace();
        return;
      }
      System.out
          .println(
              metadata.topic() + ":::partition" + metadata.partition() + "/offeset" + metadata.offset() + "/timestamp"
                  + metadata.timestamp());
    };

    this.producer.send(record, callback).get();
  }

  @Override
  public void close() {
    this.producer.close();
  }

  private static Properties properties() {
    Properties properties = new Properties();
    properties.setProperty(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "192.168.1.25:9092");
    properties.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
    properties.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, GsonSerializer.class.getName());
    return properties;
  }
}

File: simulator/project_python/kafka_dispatcher.py
from kafka import KafkaProducer


class KafkaDispatcher:
    producer = None

    def __init__(self):
        self.producer = KafkaProducer(
            bootstrap_servers=self.properties()["servers"],
            key_serializer=self.properties()["key"],
            value_serializer=self.properties()["value"],
            api_version=self.properties()["version"],
        )

    def send(self, topic, key, value):
        future = self.producer.send(topic, key=key, value=value)
        result = future.get(timeout=60)
        self.on_sucess(result)

    def on_sucess(self, record):
        print("Topic => {}".format(record.topic))
        print("Partition => {}".format(record.partition))
        print("Offset => {}".format(record.offset))
        print("TimeStamp => {}".format(record.timestamp))

    def properties(self):
        properties = {}
        properties["servers"] = ["localhost:9092"]
        properties["key"] = str.encode
        properties["value"] = str.encode
        properties["version"] = (1, 0, 0)
        return properties

    def close(self, e):
        print(str(e))
        self.producer.close()
